Cognoma ml - RandomForestClassifer - For review/advice

mans2singh opened the pull request:
Hey Folks:

I have taken a stab random forest classification algorithm implementation. I've commented out some of the SGD related coeff from the code. Please let me know if there is any advice/recommendation to improve it.

Thanks

mans2singh committed:
First cut of RandomForestClassifer

mans2singh committed:
Updated random forest params and commented out sgd coeff

mans2singh committed:
Removed duplicate params

dhimmel commented:
Nice work. Can you export your notebook to a .py file for commenting? You can do this by running the following command from the `algorithms` directory:

``` sh
jupyter nbconvert --to=script --FilesWriter.build_directory=scripts RandomForestClassifier-mans2singh.ipynb
```

mans2singh committed:
Updated grid params and added converted script

mans2singh commented:
@dhimmel - I've converted the file with the command you provided. I think there is still room to tune params. Please let me know your thoughts/recommendations.  Thanks

yl565 commented:
Nice work. I like the graphs showing how AUROC increased with increasing depth. I'm curious how the RF performs when more features are included. The current `n_feature_kept = 500` in `1.TCGA-MLexample.ipynb` is for fast demonstration not for optimal classification. I think a lot of interesting things may happen if you increase it to 5000 or even more.

mans2singh committed:
Increased sample size and estimators

mans2singh commented:
@yl565 - Thanks for your recommendation.  I've increased the number of samples and also estimators.  It does look like it has improved the performance. 
Please let me know if there is anything else required.
Thanks for your help.

gwaybio commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,272 @@
+
+# In[218]:
+
+clf = RandomForestClassifier(min_samples_leaf=5, random_state=2)
```
I think you want to specify `class_weight='balanced'` [here](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html).

gwaybio commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,272 @@
+
+# In[228]:
+
+#RandomForestClassifier does not have coefficients.
```
What about `clf.feature_importances_`?

mans2singh committed:
Updated based on comments

mans2singh commented:
@gwaygenomics - I've updated the code based on your comments (added class_weight and used clf.feature_importance_ for coeffs).  Please let me know if you have any other suggestions.

Thanks.

dhimmel commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,299 @@
+
+# In[237]:
+
+'{:.1%} zero coefficients; {:,} negative and {:,} positive coefficients'.format(
```
Since feature_importance is always positive, this cell can be removed.

Instead it may be nice to show the `feature_importance` distribution. Check out `seaborn.distplot`.

dhimmel commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,299 @@
+# In[236]:
+
+feature_importances_df = pd.DataFrame(best_clf.feature_importances_.transpose(), index=X.columns[feature_mask], columns=['weight'])
+feature_importances_df['abs'] = feature_importances_df['weight'].abs()
```
Not necessary since feature_importance is always positive.

dhimmel commented:
Nice analysis @mans2singh, just had the few comments I made above.

mans2singh committed:
Updated to remove negative cell

mans2singh commented:
@dhimmel - I've removed the negative value cells.  However, I am not sure how to use seaborn distplot.  If you have any pointer, please let me know.

Thanks

dhimmel commented:
Did you see the documentation for [`seaborn.displot`](https://stanford.edu/~mwaskom/software/seaborn/generated/seaborn.distplot.html)?

mans2singh commented:
@dhimmel  - I will check it out.  Thanks

yl565 commented:
Try also grid search min_samples_split. You could compare your performance with the Adaboost:
https://github.com/cognoma/machine-learning/pull/40#issuecomment-243418404

mans2singh commented:
Hey folks - I am getting the following exception on using seaborn plots.  This used to work earlier and I've done some google search but have not been able to resolve this issue. If anyone has any suggestion, please let me know.

[cc @dhimmel , @yl565 ]
Thanks

```
<homedirectory>anaconda/envs/cognoma-machine-learning/lib/python3.5/site-packages/matplotlib/backends/backend_agg.py in print_png(self, filename_or_obj, *args, **kwargs)
    535             close = False
    536         try:
--> 537             _png.write_png(renderer._renderer, filename_or_obj, self.figure.dpi)
    538         finally:
    539             if close:

RuntimeError: libpng signaled error

<matplotlib.figure.Figure at 0x1c6185be0>
```

dhimmel commented:
Error seems to occur [on this line](https://github.com/matplotlib/matplotlib/blob/5f553bc2f572252b00d39e687e7e194a96316909/lib/matplotlib/backends/backend_agg.py#L556) (although for a slightly earlier version of the matplotlib codebase).

Not sure what's going on. Are you using the `cognoma-machine-learning` environment? Maybe try reinstalling `libpng` within the environment by running: `conda install --force libpng`.

Also @mans2singh, note how I updated your previous post to add tripple backticks to put the error message in a code block.

mans2singh commented:
@dhimmel - I will try your recommendation. Thanks.

mans2singh committed:
Added distplot and added min_samples_split parameter for cv grid

mans2singh commented:
@dhimmel - Your recommendation to install libpng worked.  I've also added the distplot which is highly skewed though.  @yl565 - I've also added min_samples_split param to cv grid and increased features to 5000.  

Please let me know if you have any other recommendations.  

Thanks for your feedback and help.

dhimmel commented:
Nice, looks good to me. I'll let @yl565 do the merging (I recommend squashing) in case he has any additional comments.

yl565 commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,334 @@
+cv_score_mat = pd.pivot_table(cv_score_df, values='score', index='max_depth', columns='min_samples_split')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('(n_estimators)')
+ax.set_ylabel('(max_depth)');
```
Change label accordingly

yl565 commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,334 @@
+cv_score_mat = pd.pivot_table(cv_score_df, values='score', index='n_estimators', columns='min_samples_split')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('(n_estimators)')
+ax.set_ylabel('(max_depth)');
```
Same here

yl565 commented:
Overall this looks good. As expected greater tree depth gives better results as random forest usually benefit from larger trees with low bias. You should include `None` (represent no limit) as a value of `max_depth` to be grid searched otherwise the `min_samples_split` of range (2, 5) will not give different results . We have thousands of samples, try something like `min_samples_split`: [2, 10, 50, 250]

mans2singh committed:
Update labels and min sample split size

mans2singh commented:
@yl565 - I've updated the notebook based on your comments.  Please let me know if you have further suggestions.

yl565 commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,334 @@
+# Cross-validated performance heatmap
+cv_score_mat = pd.pivot_table(cv_score_df, values='score', index='max_depth', columns='n_estimators')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('(max_depth)')
```
label is still not correct. switch x and y

yl565 commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,334 @@
+cv_score_mat = pd.pivot_table(cv_score_df, values='score', index='n_estimators', columns='min_samples_split')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('(n_estimators)')
+ax.set_ylabel('(min_samples_split)');
```
same here

yl565 commented on algorithms/scripts/RandomForestClassifier-mans2singh.py:
```diff
@@ -0,0 +1,334 @@
+cv_score_mat = pd.pivot_table(cv_score_df, values='score', index='max_depth', columns='min_samples_split')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('(max_depth)')
+ax.set_ylabel('(min_samples_split)');
```
same here

yl565 commented:
Use max_depth = [1, 2, 3, 4, None]. I would expect `None` gives better results for it allows creating fully grown trees. You may use a very large number like 1000 instead `None` to facilitate results visualization.

mans2singh committed:
Corrected axis names in heat maps and max depth to 1000

mans2singh commented:
@yl565 - It was my oversight.  I've changed max_depth to 1,10,100,1000 and corrected the axis labels.  Please let me know if you see any other issue.
Thanks

yl565 commented:
Good work. Merged.
